### PR TITLE
Bundle Set up

### PR DESCRIPTION
**Pull request by WebMamba**

### CODE REVIEW

File: README.md
Comment (bocharsky-bw) [inline README.md:6]:
```suggestion
- Automatically downloads the correct Sass binary
```

File: README.md
Comment (bocharsky-bw) [inline README.md:7]:
```suggestion
- Adds a `sass:build` command to build and watch your Sass changes
```

File: README.md
Comment (bocharsky-bw) [inline README.md:19]:
```suggestion
To use generated CSS, start by including the input file
```

File: README.md
Comment (bocharsky-bw) [inline README.md:14]:
```suggestion
composer require webmamba/sass-bundle
```

File: README.md
Comment (bocharsky-bw) [inline README.md:36]:
```suggestion
php bin/console sass:build --watch
```

File: README.md
Comment (bocharsky-bw) [inline README.md:23]:
```suggestion
{# templates/base.html.twig #}
```

File: README.md
Comment (bocharsky-bw) [inline README.md:30]:
```suggestion
Then you can create your SCSS file (`assets/styles/app.scss` by default)
```

File: README.md
Comment (bocharsky-bw) [inline README.md:31]:
```suggestion
And start writing your styles.
```

File: README.md
Comment (bocharsky-bw) [inline README.md:39]:
```suggestion
And that's it!
```

File: src/Command/SassBuildCommand.php
Comment (bocharsky-bw) [inline src/Command/SassBuildCommand.php:14]:
```suggestion
    description: 'Builds the Sass assets'
```

File: src/DependencyInjection/WebMambaSassExtension.php
Comment (bocharsky-bw) [inline src/DependencyInjection/WebMambaSassExtension.php:49]:
```suggestion
                    ->info('Path to your Sass root file')
```

File: src/DependencyInjection/WebMambaSassExtension.php
Comment (bocharsky-bw) [inline src/DependencyInjection/WebMambaSassExtension.php:53]:
```suggestion
                    ->info('Path to the CSS output')
```

File: src/DependencyInjection/WebMambaSassExtension.php
Comment (bocharsky-bw) [inline src/DependencyInjection/WebMambaSassExtension.php:57]:
```suggestion
                    ->info('The Sass binary to use')
```

**Review by bocharsky-bw — COMMENTED** [review]
Some minor text tweaks :)

File: .phpunit.result.cache
Comment (weaverryan) [inline .phpunit.result.cache:1]: Don't forget to remove this :)

File: README.md
Comment (weaverryan) [reply README.md:30]:
```suggestion
Then create your SCSS file (`assets/styles/app.scss` by default)
```

File: README.md
Comment (weaverryan) [inline README.md:26]: I have a proposed change. I also wonder what @bocharsky-bw thinks about this.

Right now, the bundle works by reading `assets/styles/app.scss` and outputting `assets/styles/app.css`. You should then ignore this file from git. We can get fancier. Instead:

A) The user would reference the `app.scss` file directly - `href="{{ asset('styles/app.scss') }}"`
B) Like with TailwindBundle, we would output the Sass process to an internal, temporary file - e.g. `var/sass/app.built.css`
C) Like with TailwindBundle, we would "swap" the contents of the `styles/app.scss` file in AssetMapper for the contents of this `var/sass/app.built.css` file.

The only extra step that this bundle would need that TailwindBundle did not need is:

D) "Remap" the output filename of `styles/app.scss` in AssetMapper. If you do A-C, you will end up with a final file called something like `public/assets/styles/app.12345abcdef.scss` - ending in `.scss`, which we do not want. By decorating the core [PublicAssetsPathResolverInterface](https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/AssetMapper/Path/PublicAssetsPathResolverInterface.php), we can change the output name of this file to end in `.css`.

**Review by weaverryan — COMMENTED** [review]
TYVM for getting this going! I dropped some comments

File: README.md
Comment (bocharsky-bw) [reply README.md:26]: Hm, on the one hand, linking to a CSS file is a more straightforward way... but as you said that file should be probably "gitignored", so I think it really better to link directly to the Scss file then excluding working with auto-generated file directly. And I like to make the behavior consistent with TailwindBundle  👍

**Comment by WebMamba** [comment]
Thanks for your review @weaverryan and @bocharsky-bw. I fix the syntax issues you point out, rename webmamba to symfonycasts, and now the user reference directly app.scss as in the last comment. Tell me what you thinks about all this changes! Cheers 😁🧡

File: README.md
Comment (weaverryan) [inline README.md:20]: We should first tell them to create an `assets/styles/app.scss` file - even maybe give some very basic example file - e.g. one that sets a variable, then uses that to set the `background-color` of the `body` element.

Then it will make more sense whey we're pointing to `styles/app.scss` below. Also, type on the line above - it says `app.css` :)

File: README.md
Comment (weaverryan) [inline README.md:31]: This sentence will now be moved above (per my comment a few lines up). Replace it with something like:

```rst
That's right! You point directly to the `.scss` file. But don't worry, the final built `.css` file will be returned!
```

File: README.md
Comment (weaverryan) [inline README.md:39]: We should include a few more sections - like we do over here - https://github.com/SymfonyCasts/tailwind-bundle#how-does-it-work

File: composer.json
Comment (weaverryan) [inline composer.json:25]: Let's use `phpunit-bridge`, if nothing else for consistency with our other libs - https://github.com/SymfonyCasts/tailwind-bundle/blob/main/composer.json#L23

File: src/AssetMapper/SassCssCompiler.php
Comment (weaverryan) [inline src/AssetMapper/SassCssCompiler.php:23]: Actually, this should be the final, output CSS file - so `$this->cssPath` I believe. What we're saying here is:

> Hey AssetMapper! When the final, built CSS file changes, you need to bust the cache on this asset and serve the fresh contents.

File: src/AssetMapper/SassCssCompiler.php
Comment (weaverryan) [inline src/AssetMapper/SassCssCompiler.php:9]: Can you setup php-cs-fixer - like https://github.com/SymfonyCasts/tailwind-bundle/blob/main/.php-cs-fixer.dist.php - but tweak the library name in the "header"? That'll add the license header to each file.

File: src/AssetMapper/SassPublicPathAssetPathResolver.php
Comment (weaverryan) [inline src/AssetMapper/SassPublicPathAssetPathResolver.php:16]: Ideally we would target JUST the one `.scss` file. Since we are only given the logical path here... and the `root_sass` config is an absolute path, that makes it tricky.

WDYT about changing the `root_sass` to be the asset mapper "logical path" instead? Then it would be easy to compare that logical path to this logical path. And, it makes sense: your "input Sass" file MUST be in an AssetMapper path for all of this to work. If we decide to make that change and it works, I'll do the same for tailwind-bundle.

**Review by weaverryan — COMMENTED** [review]
Ran out of time to fully review - but I have some to keep you moving! More coming hopefully tomorrow. Looking good - excited to get this done.

File: composer.json
Comment (weaverryan) [inline composer.json:14]: was there ultimately an extension needed for unzipping the tar?

File: tests/fixtures/assets/app.scss
Comment (weaverryan) [inline tests/fixtures/assets/app.scss:3]: Is this meant to be here? I'm asking because `FunctionalTest` seems to create this in its `setUp()`.

Also, if we set a variable then use it, the assertion in the test will be a bit more meaningful, as we'll see Sass doing its job.

**Review by weaverryan — CHANGES_REQUESTED** [review]
A few more comments - now fully reviewed. Thanks :)

File: src/SassBinary.php
Comment (bendavies) [inline src/SassBinary.php:12]: how about if this was configurable by the user?
that way, the user can upgrade sass, without waiting for a new release of the bundle.

Comment (weaverryan) [reply src/SassBinary.php:12]: Seems reasonable to me - the constant would be the default. It wouldn't surprise me also if we got a request in the future where someone needed to "lock" to an older version.

**Comment by bendavies** [comment]
I think i have a feature request to be considered before this is merged.
The request is to support compiling/watching more than one scss file.
This is required to support [page specific css](https://symfony.com/doc/current/frontend/asset_mapper.html#page-specific-css-javascript), and to migrate effectively from "multiple entrypoints" in webpack encore.

this is supported fine by the cli:
https://sass-lang.com/documentation/cli/dart-sass/#many-to-many-mode


this would require changing the `root_sass` configuration option to an array node to accept multiple paths

File: src/DependencyInjection/SymfonycastsSassExtension.php
Comment (bendavies) [reply src/DependencyInjection/SymfonycastsSassExtension.php:53]: use case provied!
https://github.com/SymfonyCasts/sass-bundle/pull/1#issuecomment-1644709073

**Comment by weaverryan** [comment]
That makes sense to me. I can’t immediately think why this would cause any major issues.

File: composer.json
Comment (WebMamba) [reply composer.json:14]: Looks like it's enable by default: https://www.php.net/manual/en/phar.installation.php

File: tests/fixtures/assets/app.scss
Comment (WebMamba) [reply tests/fixtures/assets/app.scss:3]: Yes this is for the SassBuilderTest

File: src/AssetMapper/SassPublicPathAssetPathResolver.php
Comment (WebMamba) [reply src/AssetMapper/SassPublicPathAssetPathResolver.php:16]: Hum sorry I don't get what you mean here! Maybe something I miss understand with asset mapper

**Comment by WebMamba** [comment]
@bendavies thanks for your review! I implemented the support for multiples sass files. 😁

File: src/ScssFileNameGuesser.php
Comment (bendavies) [inline src/ScssFileNameGuesser.php:11]: i think this should just be `$filename.css` so it aligns with the assetmapper docs nicely

**Comment by bendavies** [comment]
> @bendavies thanks for your review! I implemented the support for multiples sass files. 😁

great! thanks

File: src/SassBinary.php
Comment (weaverryan) [inline src/SassBinary.php:92]: This fails because, by default, `$this->binaryPath` is null. This should actually reuse the logic from `createProcess()` where it uses `$this->binaryDownloadDir.'/dart-sass/sass';`.

Comment (weaverryan) [reply src/SassBinary.php:92]: Also, we should move the binary into `var/sass` instead of `bin/...` as there is no reason for the user to wonder whether they should commit this file or not. This is how it's done in TailwindBundle - https://github.com/SymfonyCasts/tailwind-bundle/blob/main/config/services.php#L20C17-L20C22 - that's eventually passed to `new TailwindBinary()`: https://github.com/SymfonyCasts/tailwind-bundle/blob/main/src/TailwindBuilder.php#L101C42-L101C56

File: src/SassBuilder.php
Comment (weaverryan) [inline src/SassBuilder.php:69]:
```php
if (!\is_file($sassPath)) {
    throw new \Exception(sprintf('Could not find Sass file: "%s"', $sassPath));
}
```

File: tests/fixtures/bin/dart-sass/sass
Comment (weaverryan) [inline tests/fixtures/bin/dart-sass/sass:6]: Can we delete this directory? It'd be better (and I believe this is what we do in TailwindBundle) if the test actually downloaded the binary it needs to test the whole process (and then we can test on Windows, for example, in the CI).

File: src/SassBuilder.php
Comment (weaverryan) [inline src/SassBuilder.php:70]: Minor: let's inline the `guessCssNameFromSassFile()` into this class, make it `public` (non-static, see below) with`@internal` on it. Call the method: `getSassOutputFilename()`. 

Then, we inject `SassBuilder` into the compiler class. This will avoid needing to pass the scss paths around to multiple classes: make `SassBuilder` the central class that's aware of this, and give it any public methods it needs to fetch off that info. It just seems a bit simpler - we do it in TailwindBundle - https://github.com/SymfonyCasts/tailwind-bundle/blob/main/src/AssetMapper/TailwindCssAssetCompiler.php#L28

**Review by weaverryan — COMMENTED** [review]
Left some comments - but this is working really nicely now! After these comments are addressed, we should be able to merge.

Things still needed after that:

A) Adding CI
B) #2 (not a blocker for an initial release, but this would be REALLY good to have)
C) Thinking about a shared "build" bundle that Tailwind & Sass could both use - the idea being that someone could, in theory, use Tailwind AND Sass and use some central `assets:build` command that would take the source contents through a small pipeline (e.g. run Sass on a file, take its output, run Tailwind on that to get the final contents).

File: src/SassBinary.php
Comment (bendavies) [reply src/SassBinary.php:92]: I'm not sure how I feel about that. This pretty much breaks all existing conventions of having binaries in `bin`.
How about keeping in `bin` and adding a recipe to add a git ignore?

**Comment by WebMamba** [comment]
Great! Lets have a final check, all your comment should be fixed now! 😁

File: src/SassBinary.php
Comment (weaverryan) [reply src/SassBinary.php:92]: The binary is a completely hidden thing - the user would never need to run it directly (and won't commit it). My thoughts: why put it in `bin/` for the user to see and worry about?

**Comment by weaverryan** [comment]
This is working fantastically! Thank you a million @WebMamba!

**Pull request merged**
